SIPS/sip-11.md

```diff
@@ -0,0 +1,62 @@
+---
+sip: 11
+title: Transaction Insights V2 - Transaction Warnings
+status: Draft
+discussions-to: https://github.com/MetaMask/SIPs/discussions/101
+author: Hassan Malik (@hmalik88)
+created: 2023-07-16
+---
+
+## Abstract
+
+This SIP proposes a way for snaps to provide extra friction for executing potentially dangerous transactions within MetaMask.
+
+## Motivation
+
+One of the biggest issues with wallet users is a loss of funds through executing some sketchy or seemingly innocuous transaction. With the benefit of the transaction insights API, a snap can already run analysis on an unsigned transaction paylod. Why not allow for the snap to provide extra friction to the user if it determines that a transaction is malicious? Users often click through things without reading, by allowing for a warning we can add friction at points that we really think a user should have a second glance.
+
+This warning could be injected at any point where there are insights provided.
+
+## Specification
+
+Please see [SIP-3](https://github.com/MetaMask/SIPs/blob/main/SIPS/sip-3.md) for more information on the original transaction insights API.
+
+### Language
+
+The key words "MUST", "MUST NOT", "REQUIRED", "SHALL", "SHALL NOT",
+"SHOULD", "SHOULD NOT", "RECOMMENDED", "NOT RECOMMENDED", "MAY", and
+"OPTIONAL" written in uppercase in this document are to be interpreted as described in [RFC 2119](https://www.ietf.org/rfc/rfc2119.txt)
+
+### Snap Implementation
+
+Any snap that wishes to provide transaction insight features **MUST** implement the following API:
+
+```typescript
+import { OnTransactionHandler } from "@metamask/snap-types";
+
+export const onTransaction: OnTransactionHandler = async ({
+  transaction,
+  chainId,
+}) => {
+  const insights = /* Get insights */;
+  const willFail = () => /* Return boolean */;
+  return willFail(insights) ? { insights, warning: 'Some warning message' } : { insights };
+};
+```
+
+The interface for the return value of an `onTransaction` export is:
+
+```typescript
+interface OnTransactionReturn {
+  insights: Record<string, Json>;
+  warning?: string;
```

Comment:
I thought this was supposed to be a flag or an enum value instead of a string?

Are we not just showing the insights content again in the modal?

cc @eriknson

Reply:
Oh maybe? I was just going off the designs. Haven't had a chance to speak to Erik yet.

Reply:
Oh ye, it's the same insight. And it'll only populate the modal if it's in critical / warning state (boolean)

Reply:
I think we should have this be an enum instead of a boolean value, for now we can just leave the enum with two states: "none" or "critical". This makes it easy to expand to add more states if we ever want to.